### DISCUSSION TITLE
Point constraint limitation

### DISCUSSION THREAD

**Discussion by tksuoran** [jrouwe/JoltPhysics#268] [Q&A] [answered]
Bodies do not seem to gain kinetic energy when they are moved by point constraint.
Can I somehow make a point constraint that behaves more physically?
I think I would need a point constraint with motor. However, point constraint does not seem to support motor.
Can I use 6DOF constraint as is somehow for this? Or would this need some more customization?

**Using Jolt point constraint to drag bodies:**
https://user-images.githubusercontent.com/994606/195119041-e6e79753-3521-44a4-9522-3c9cff17d984.mp4
- Bodies do not seem to gain any kinetic energy when dragged with point constraint
- Dragged bodies do not pass (much) kinetic energy to other bodies they interact with

**Using Bullet point constraint to drag bodies:**
https://user-images.githubusercontent.com/994606/195119092-c12988ec-8336-4da5-adbf-be1332fa5c96.mp4
- Bodies gain velocity when dragged
- Dragged bodies behave physically when interacting with bodies

**Comment by jrouwe** [comment, marked as answer]
A point constraint has no spring element in it, all it does it to try to keep the points together. If all you do is to set the world attachment point every frame, then the velocity of that point will be zero (because you've attached it to the fixed world). Constraints are solved in the velocity domain (by SolveVelocityConstraint) and by setting a new position you haven't violated the relative velocity between the bodies. The position of the body is incorrect though, but this will be fixed up by SolvePositionConstraint which will just fix up the positions and not actually introduce any velocity on the body. This works differently in Bullet, where the position error is solved together with the velocity error, which effectively gives you the spring you're looking for (but it has the disadvantage that it adds energy to the system and can cause oscillations).

Long story short, you can fix it in 2 ways:
* Use a DistanceConstraint with a non-zero mFrequency and mDamping (see SamplesApp::UpdateDebug)
* Instead of fixing the other end of your PointConstraint to Body::sFixedToWorld, create a Kinematic body and move it along with the hand using BodyInterface::MoveKinematic (this ensures that your 'hand' has velocity that gets transferred to the body you're lifting)

**Reply by tksuoran** [reply]
I tried the second method, but it did not seem to behave any better. In my implementation, I added a new non-colliding broad phase layer for the drag end body. I tested both basic sphere shape and a custom "empty" shape which would not collide with anything, these behaved the same (in Bullet version I used btEmptyShape).

I will try the DistanceConstraint next 🤞🏻

**Reply by tksuoran** [reply]
Using `DistanceConstraint` worked - thank you very much.
